File: Data Structures & Algorithms/generate-parentheses/submission-0.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        if n == 1:
            return ["()"]
        res = []
        def back(cur, op, clo):
            if op == n and clo == n:
                res.append(cur)
                return
            if op < n:
                back(cur + "(",op + 1,clo)
            if clo < op:
                back(cur + ")",op,clo + 1)
        
        back("",0,0)
        return res

File: Data Structures & Algorithms/combinations/submission-0.cpp
class Solution {
public:
    void back(int n,int i, int k, vector<vector<int>>& res, vector<int>& sub){
        if (sub.size() == k){
            res.push_back(sub);
            return;
        }
        for(int j = i;j <= n;j++){
            sub.push_back(j);
            back(n,j + 1,k,res,sub);
            sub.pop_back();
        }
    }
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int>> res;
        res.clear();
        vector<int> sub;
        back(n,1,k,res,sub);
        return res;
    }
};